include/nbl/asset/ICPUDescriptorSet.h

```diff
@@ -210,29 +217,43 @@ class NBL_API ICPUDescriptorSet final : public IDescriptorSet<ICPUDescriptorSetL
 			{
 				--_levelsBelow;
 				restoreFromDummy_impl_call(m_layout.get(), other->getLayout(), _levelsBelow);
-				for (auto it = m_descriptors->begin(); it != m_descriptors->end(); it++)
+
+				for (uint32_t t = 0u; t < EDT_COUNT; ++t)
 				{
-					auto descriptor = it->desc.get();
-					if (!descriptor)
+					const auto type = static_cast<E_DESCRIPTOR_TYPE>(t);
+					const auto descriptorCount = m_layout->getTotalDescriptorCount(type);
+					if (descriptorCount == 0ull)
 						continue;
-					const auto i = it - m_descriptors->begin();
-					auto* d_other = other->m_descriptors->begin()[i].desc.get();
 
-					switch (descriptor->getTypeCategory())
+					auto descriptors = m_descriptors[type]->begin();
+					assert(descriptors);
+
+					auto otherDescriptors = other->m_descriptors[type]->begin();
+
+					for (uint32_t i = 0u; i < descriptorCount; ++i)
 					{
-					case IDescriptor::EC_BUFFER:
-						restoreFromDummy_impl_call(static_cast<ICPUBuffer*>(descriptor), static_cast<ICPUBuffer*>(d_other), _levelsBelow);
-						break;
-					case IDescriptor::EC_IMAGE:
-						restoreFromDummy_impl_call(static_cast<ICPUImageView*>(descriptor), static_cast<ICPUImageView*>(d_other), _levelsBelow);
-						if (descriptor->getTypeCategory() == IDescriptor::EC_IMAGE && it->image.sampler)
-							restoreFromDummy_impl_call(it->image.sampler.get(), other->m_descriptors->begin()[i].image.sampler.get(), _levelsBelow);
-						break;
-					case IDescriptor::EC_BUFFER_VIEW:
-						restoreFromDummy_impl_call(static_cast<ICPUBufferView*>(descriptor), static_cast<ICPUBufferView*>(d_other), _levelsBelow);
-						break;
+						switch (descriptors[i]->getTypeCategory())
```

Comment:
you already know the type category, shouldn't need to query it (derive, since you know `t`)

Reply:
Resolved.